docs/design/2022-11-25-global-resource-control.md

```diff
@@ -0,0 +1,240 @@
+```go
+ProtoBuf
+message TokenBucketRequest {
+    RequestHeader header = 1;
+    repeated TokenBucketConsumption comsumption_stats;
+}
+message TokenBucketResponse {
+    ResponseHeader header = 1;
+    double granted_rru = 2;
+    double granted_wru = 2;
+    int64  trickle_duration_seconds = 3;
+    double fallback_rru_rate = 4;
+    double fallback_wru_rate = 5;
+}
+message TokenBucketConsumption {
+    uint64 tag = 1; // resource group tag
+    Consumption consumption_since_last_request = 2 [(gogoproto.nullable) = false];
+    string instance_fingerprint = 3;
+    double requested_rru = 4;
+    double requested_wru = 4;
+    uint64 target_request_period_seconds = 5;
+}
+message Consumption {
+    double rru = 1;
+    double wru = 2;
+    uint64 read_requests = 3;
+    uint64 read_bytes = 4;
+    uint64 write_requests = 5;
+    uint64 write_bytes = 6;
+    double cpu_seconds = 6;
+}
+```
+
+In this protocol, there are some key fields:
+- `target_request_period_seconds`: This field is the lease interval for the local token buckets, it is configurable.
+- `requested_rru/ request_wru`: It's a load expect with load estimation. This is an exponentially weighted moving average of the load at the nod.
+
+#### Bursting and Refill rate
+Bucket Tokens have two parts.
+- `burst`: it is an accumulated value, it absorbs temporal spikes in workloads of the tenant user. Moreover, if the system is elastic enough, such a cluster can be auto-scale in the cloud. It can be used as the budget for users to pay. If the burst is enough, customers feel free to use resources without limitation, which matches the main idea of the serverless product.
+- `refill rate`: rate-limit resource consumption. grant the tokens over a specified period of time, according to a fraction of the global refill rate
+Once a request with a token bucket server is completed, the node adjusts its local token bucket according to the response. either adding burst tokens or setting up a refill rate.
+#### Debt
+The token buckets can be "debt", as we mentioned, due to some statistics only being accounted for after the fact (eg, the read size of coprocess request, the CPUSecs usage). When a local bucket is in debt, KV operations are blocked until the debt is paid. This may lead to the latency not being stable . There is an improvement that allows the debt to be paid over a time period from the refill rate.
+#### Adaptive Capacity
+When a request from the application arrives, the request deducts tokens. Eventually, the LAC will run out of tokens because of consumption or expiry. When the LAC runs out of tokens, it requests more tokens from Global Token Server immediately. The Glocal Token Server instance uses the information provided by the client to estimate the global token consumption and vends tokens available for the next time unit to the client’s share of overall tokens. Thus, it ensures that workloads can adaptively turn the  local bucket fill rage up to the maximum capacity.
```

Comment:
```suggestion
When a request from the application arrives, the request deducts tokens. Eventually, the LAC will run out of tokens because of consumption or expiry. When the LAC runs out of tokens, it requests more tokens from Global Token Server immediately. The Glocal Token Server instance uses the information provided by the client to estimate the global token consumption and vends tokens available for the next time unit to the client's share of overall tokens. Thus, it ensures that workloads can adaptively turn the local bucket fill rage up to the maximum capacity.
```